cascade/dynamics.py

```diff
@@ -0,0 +1,127 @@
+        self.processes = processes.copy()
+        self.scratch_dir = scratch_dir
+
+    def run_dynamics(self,
+                     start: Progress,
+                     model_msg: bytes | State,
+                     learner: BaseLearnableForcefield,
+                     max_timesteps: int,
+                     max_frames: int | None = None,
+                     device: str = None) -> tuple[bool, list[Atoms]]:
+        """Run dynamics for a maximum number of timesteps using a particular forcefield
+
+        Runs dynamics until the end of a process or until the maximum number of timesteps is reached.
+
+        Args:
+            start: Starting point of the dynamic trajectory
+            model_msg: Serialized form of the forcefield used for training
+            learner: Class used to generate the ASE calculator object
+            max_timesteps: Maximum number of timesteps to run dynamics
+            max_frames: Maximum number of frames from the atomistic simulation to return for auditing
+            device: Device to use for evaluating forcefield
+        Returns:
+            List of frames selected at used for auditing the dynamics
+        """
+
+        # Create a temporary directory in which to run the data
+        dyn_cls, dyn_steps, dyn_args, run_args, post_func = self.processes[start.process]  # Pick the current process
+        with TemporaryDirectory(dir=self.scratch_dir, prefix='cascade-dyn_', suffix=f'_{start.name}') as tmp:
+            tmp = Path(tmp)
+            dyn = dyn_cls(start.atoms, logfile=str(tmp / 'dyn.log'), **dyn_args)
+
+            # Attach the calculator
+            calc = learner.make_calculator(model_msg, device)
+            atoms = start.atoms
+            atoms.calc = calc
+
+            # Attach the trajectory writer
+            traj_freq = 1 if max_frames is None else max_timesteps // max_frames
+            traj_path = str(tmp / 'run.traj')
+            with Trajectory(traj_path, mode='w', atoms=start.atoms) as traj:
+                dyn.attach(traj, traj_freq)
+
+                # Run dynamics, then check if we have finished
+                converged = dyn.run(steps=max_timesteps, **run_args)
+                total_timesteps = max_timesteps + start.timestep  # Total progress along this step
+
+                if converged and isinstance(dyn, Optimizer):  # Optimization is done if convergence is met
+                    done = True
+                elif isinstance(dyn, Optimizer) and dyn_steps is not None:  # Optimization is also done if we've run out of timesteps
+                    done = total_timesteps >= dyn_steps
+                else:
+                    done = total_timesteps >= dyn_steps
+
+                if done and post_func is not None:
+                    post_func(atoms)
+
+            # Read in the trajectory then append the current frame to it
+            traj_atoms = read(traj_path, ':')
+            traj_atoms.append(atoms)
```

Comment:
>  the atoms object is updated in place 
Correct

> the most recent update is not written to disk
It might not be, if the number of steps being run is not a multiple of the timestep frequency. So, I append the current structure to the trajectory to be sure.

I should probably verify I'm not duplicating it